File: 19-remove-nth-node-from-end-of-list/remove-nth-node-from-end-of-list.cpp
/**
 * Definition for singly-linked list.
 * struct ListNode {
 *     int val;
 *     ListNode *next;
 *     ListNode() : val(0), next(nullptr) {}
 *     ListNode(int x) : val(x), next(nullptr) {}
 *     ListNode(int x, ListNode *next) : val(x), next(next) {}
 * };
 */
class Solution {
public:
    ListNode* removeNthFromEnd(ListNode* head, int n) {
        ListNode* temp = head;
        while(n>0){
            temp = temp->next;
            n--;
        }
        if(temp==NULL) return head->next;// we need to remove head
        ListNode* t1=head;
        ListNode* prev;
        while(temp!=NULL){
           temp = temp->next;
           prev = t1;
           t1=t1->next;
        }
        prev->next = t1->next;
        return head;
    }
};

File: 19-remove-nth-node-from-end-of-list/remove-nth-node-from-end-of-list.java
/**
 * Definition for singly-linked list.
 * public class ListNode {
 *     int val;
 *     ListNode next;
 *     ListNode() {}
 *     ListNode(int val) { this.val = val; }
 *     ListNode(int val, ListNode next) { this.val = val; this.next = next; }
 * }
 */
class Solution {
    public ListNode removeNthFromEnd(ListNode head, int n) {
        if(head==null) return head;
        ListNode t1 = head;
        while(n>0)
        {
            t1 = t1.next;
            n--;
        }
        if(t1==null) return head.next;
        ListNode t2 = head;
        ListNode prev = null;
        while(t1!=null){
            prev = t2;
            t1 = t1.next;
            t2 = t2.next;
        }
        if(t2.next!=null) {
            prev.next = t2.next;
        }
        else prev.next = null;

        return head;

        
    }
}